servers/audio/effects/audio_effect_capture.cpp

```diff
@@ -0,0 +1,158 @@
+	}
+
+	PackedVector2Array ret;
+	ret.resize(p_frames);
+
+	Vector<AudioFrame> streaming_data;
+	streaming_data.resize(p_frames);
+	buffer.read(streaming_data.ptrw(), p_frames);
+	for (int32_t i = 0; i < p_frames; i++) {
+		ret.write[i] = Vector2(streaming_data[i].l, streaming_data[i].r);
+	}
+	return ret;
+}
+
+void AudioEffectCapture::clear_buffer() {
+	const int32_t data_left = buffer.data_left();
+	buffer.advance_read(data_left);
+}
+
+void AudioEffectCapture::_bind_methods() {
+	ClassDB::bind_method(D_METHOD("can_get_buffer", "frames"), &AudioEffectCapture::can_get_buffer);
+	ClassDB::bind_method(D_METHOD("get_buffer", "frames"), &AudioEffectCapture::get_buffer);
+	ClassDB::bind_method(D_METHOD("clear_buffer"), &AudioEffectCapture::clear_buffer);
+	ClassDB::bind_method(D_METHOD("set_buffer_length", "buffer_length_seconds"), &AudioEffectCapture::set_buffer_length);
+	ClassDB::bind_method(D_METHOD("get_buffer_length"), &AudioEffectCapture::get_buffer_length);
+	ClassDB::bind_method(D_METHOD("get_frames_available"), &AudioEffectCapture::get_frames_available);
+	ClassDB::bind_method(D_METHOD("get_discarded_frames"), &AudioEffectCapture::get_discarded_frames);
+	ClassDB::bind_method(D_METHOD("get_buffer_length_frames"), &AudioEffectCapture::get_buffer_length_frames);
+	ClassDB::bind_method(D_METHOD("get_pushed_frames"), &AudioEffectCapture::get_pushed_frames);
+
+	ADD_PROPERTY(PropertyInfo(Variant::FLOAT, "buffer_length", PROPERTY_HINT_RANGE, "0.01,10,0.01"), "set_buffer_length", "get_buffer_length");
+}
+
+Ref<AudioEffectInstance> AudioEffectCapture::instance() {
+	if (!buffer_initialized) {
+		int64_t buffer_max_size = (int64_t)(1000.0f * buffer_length_seconds);
+		buffer_max_size *= AudioServer::get_singleton()->get_mix_rate();
+		buffer_max_size /= 1000; //convert to seconds
+		ERR_FAIL_COND_V(buffer_max_size <= 0 || buffer_max_size >= (1 << 27), Ref<AudioEffectInstance>());
+		buffer.resize(nearest_shift((int32_t)buffer_max_size));
+		buffer_initialized = true;
+	}
+
+	clear_buffer();
+
+	Ref<AudioEffectCaptureInstance> ins;
+	ins.instance();
+	ins->base = Ref<AudioEffectCapture>(this);
+
+	return ins;
+}
+
+void AudioEffectCapture::set_buffer_length(float p_buffer_length_seconds) {
+	ERR_FAIL_COND(buffer_initialized);
+
+	buffer_length_seconds = p_buffer_length_seconds;
+}
+
+float AudioEffectCapture::get_buffer_length() {
+	return buffer_length_seconds;
+}
+
+int AudioEffectCapture::get_frames_available() const {
+	ERR_FAIL_COND_V(!buffer_initialized, 0);
+	return buffer.data_left();
+}
+
+int64_t AudioEffectCapture::get_discarded_frames() const {
+	return discarded_frames;
+}
+
+int AudioEffectCapture::get_buffer_length_frames() const {
+	ERR_FAIL_COND_V(!buffer_initialized, 0);
+	return buffer.size();
+}
+
+int64_t AudioEffectCapture::get_pushed_frames() const {
+	return pushed_frames;
+}
+
+AudioEffectCapture::AudioEffectCapture() {
+}
+
+AudioEffectCapture::~AudioEffectCapture() {
+}
+
+void AudioEffectCaptureInstance::init() {
```

Comment:
This doesn't seem to be used, is it overriding a virtual method? For empty overrides and constructor/destructor I'd suggest just adding the `{}` in the header.